crates/bevy_ecs/src/schedule/executor/async.rs

```diff
@@ -0,0 +1,432 @@
+    OnceLock<
+        KeyedQueues<
+            (WorldId, InternedScheduleLabel),
+            (Waker, fn(&mut World, AsyncTaskId), AsyncTaskId),
+        >,
+    >,
+);
+
+impl WakeRegistry {
+    pub fn wait(&self, schedule: InternedScheduleLabel, world: &mut World) -> Option<()> {
+        let world_id = world.id();
+        let mut waker_list = std::vec![];
+        while let Ok((waker, system_init, task_id)) = GLOBAL_WAKE_REGISTRY
+            .0
+            .get_or_init(|| KeyedQueues::new())
+            .get_or_create(&(world_id, schedule))
+            .pop()
+        {
+            // It's okay to call this every time, because it only *actually* inits the system if the task id is new
+            system_init(world, task_id);
+            waker_list.push(waker);
+        }
+        let waker_list_len = waker_list.len();
+        if waker_list_len == 0 {
+            return None;
+        }
+        world.insert_resource(WakeParkBarrier(
+            thread::current(),
+            Arc::new(AtomicI64::new(waker_list_len as i64 - 1)),
+        ));
+        if let None = GLOBAL_WORLD_ACCESS.set(world, || {
+            for waker in waker_list {
+                waker.wake();
+            }
+            thread::park();
+        }) {
+            return None;
+        }
+        world.try_resource_scope(|world, mut appliers: Mut<SystemParamAppliers>| {
+            appliers.run(world);
+        });
+        Some(())
+    }
+}
+
+/// The PhantomData here is just there cause it's a cute way of showing that we have a mutex around our unsafe worldcell and that's what the mutex is 'locking'
+pub(crate) struct WorldAccessRegistry(
+    OnceLock<
+        RwLock<
+            HashMap<
+                WorldId,
+                RwLock<
+                    Option<(
+                        UnsafeWorldCell<'static>,
+                        Mutex<PhantomData<UnsafeWorldCell<'static>>>,
+                    )>,
+                >,
+            >,
+        >,
+    >,
+);
+
+impl WorldAccessRegistry {
+    pub(crate) fn set(&self, world: &mut World, func: impl FnOnce()) -> Option<()> {
+        let this = self.0.get_or_init(|| RwLock::new(HashMap::new()));
+        let world_id = world.id();
+        if !this.read().unwrap().contains_key(&world_id) {
+            // VERY rare only happens the first time we try to do anything async in a new World
+            let _ = this.write().unwrap().insert(world_id, RwLock::new(None));
+        }
+
+        struct ClearOnDropGuard<'a> {
+            slot: &'a RwLock<
+                Option<(
+                    UnsafeWorldCell<'static>,
+                    Mutex<PhantomData<UnsafeWorldCell<'static>>>,
+                )>,
+            >,
+        }
+        impl<'a> Drop for ClearOnDropGuard<'a> {
+            fn drop(&mut self) {
+                // clear it on the way out, even on panic
+                self.slot.write().unwrap().take();
+            }
+        }
+        unsafe {
```

Comment:
This unsafe block is too large; shrink it to only the relevant operations (afaictjust the transmute?)

Reply:
Hmm I am *not* confident that's how it's done in the bevy engine, that is the broader way that i've seen it being done, but bevy might do it differently, i'll wait for someone to chime in who knows and if not i'll defer to you.